### ISSUE TITLE
Support for data incremental scenario

### ISSUE THREAD

**Issue by AntonioCarta** [ContinualAI/avalanche#360]
We need to support the data incremental scenario with no task labels, no task boundaries. The problem is that experiences always have boundaries between them. After a bit of discussion I think there are two possible solutions:

1 - **single experience**, where the entire scenario is a single `Experience` carefully loaded by a data loader that does not reuse data.
2 - **a separate experience for each mini-batch**, where boundaries happen at the mini-batch level, which makes each mini-batch independent.

Personally, I prefer option (2). It seems natural to me to represent the passage of time using the stream of experiences. When you have a stream of experiences, it's clear that you either store them in a buffer or you will not encounter anymore. Option (1) is possible, and maybe slightly more efficient, but it's more error-prone. If you don't use a proper data loader it's not data incremental anymore. If you have multiple epoch it's not data incremental. And so on.

@lrzpellegrini what do you think? Is it possible to have such a large number of experiences?

@Mattdl what do you think of this solution?

See also [previous discussion](https://github.com/ContinualAI/avalanche/issues/336#issuecomment-789172633).

**Comment by Mattdl** [comment]
That's a great solution @AntonioCarta , I would also prefer option 2. The problem is though that right now the 'experiences' of the dataset creation and the 'experience' in the Strategy flow are entangled. We should still be able to create a datastream in a set of steps (or experiences, such as e.g. typical Split-MNIST), while the algorithm does batch-wise processing (if the batch would be an experience, this becomes a problem). It should also be possible to process the same batch multiple times as is often done in the online data incremental scenarios.

**Comment by vlomonaco** [comment]
I think we already support (2) it's just that it is probably not that efficient, but I'll let @lrzpellegrini brief us on this.

**Comment by AntonioCarta** [comment]
> The problem is though that right now the 'experiences' of the dataset creation and the 'experience' in the Strategy flow are entangled. 

Sorry, what do you mean?

> It should also be possible to process the same batch multiple times as is often done in the online data incremental scenarios.

This is already possible, and it works also in the data incremental scenario with option (2).

**Comment by lrzpellegrini** [comment]
Having an experience for each mini-batch (2) is totally doable. It shouldn't add too much overhead.

A custom scenario can already generate the required experiences if correctly implemented, even experiences containing 1 single pattern. Also, experiences are now created on the fly in a lazy way and the underlying dataset is generated based on lists of indexes, so this is why expect little to no overhead.

The point is, do we need some other way to create mini-batch experiences out of initial experiences?

- It should be doable to add a "split" method to the `Experience` class to allow the creation of sub-experiences of a certain size (or based on other criteria) on the fly (maybe in the very first step of the strategy). However, in that case we would need to rethink the `current_experience` field in the `Experience` class. Also, this would really entangle the scenario definition with the strategy.
- Another solution would be implementing the "split" method at a stream level, which would allow the stream to output pre-split experiences, thus changing the semantic of the scenario itself (which, if I got the point of the discussion, is the right thing to do). I think that this solution is better because 1) The definition of the scenario would be completely handled outside the strategy, possibly inside the scenario generator as it should be; 2) The `current_experience` field of the `Experience` class would make sense.

I'm not sure that (1) would be a great idea: it makes things complicated in my opinion. The good thing about having experiences is that they clearly define "a bunch of things you won't able to see again"... I don't really like idea that the order of things should be managed at the DataLoader level.

**Comment by AntonioCarta** [comment]
> The point is, do we need some other way to create mini-batch experiences out of initial experiences?

I don't think we need this if we already support data incremental. 

Since you agree that there are no limitations for creating a large number of experiences (especially in a lazy way, which may be useful), I think we only need to provide some examples of scenarios already implemented and the documentation at all levels (examples, tutorials, api). The main problem that I see right now is that most of the end users may not clearly see how avalanche streams map to the common benchmarks in the literature.

> The good thing about having experiences is that they clearly define "a bunch of things you won't able to see again"

Totally agree. This is another point that we should stress in the documentation.

**Comment by Mattdl** [comment]
> > The problem is though that right now the 'experiences' of the dataset creation and the 'experience' in the Strategy flow are entangled.
> 
> Sorry, what do you mean?

The problem is that now:
1) We build a data stream made up of a list of experiences (e.g. tasks). Here the experience defines how to subdivide the stream in subsets.
2) However, if we use 1 experience per batch (proposed solution above), this experience also defines the unit of processing. That means we have to subdivide the stream in a lot of batch-sized experiences, but it should be possible to still divide the data stream in just e.g. task-based experiences. (For example, for data incremental works to be able to compare to class-incremental works).

TLDR: We should be able to define a datastream subdivided in tasks (experiences), without constraining the unit of processing (e.g. we shouldn't need to enforce batch-sized experiences)

Hope this is more clear (:
